### PR TITLE
1. Restore documents to correct position to begin with instead of ope…

### PR DESCRIPTION
**Pull request by gmcode**
…nning at start then moving to correct position. This prevents significant flickering and should be quicker. Remove wrapping work around code that seems uneccessary. 2. Try to address perceived issues with backup/restore in event of file existence errors. 3. Don't activate current tab twice. Please test all of this very carefully, not to be rushed into release.

This is a significant change to address the main document painting issues I have with bowpad - which is that it paints the first screen of a document and then repaints the screen again with the actual position it wants to restore to. If you use MainWindow.cpp as an example, by going to the end of the document and restarting, you should see the comments section of the file appear then vanish as bowpad repaints the end section. This patch aims to fix that problem.

The changes in this patch touch the cursor handling and backup/restore stuff so it needs careful testing as I'm not claiming to really understand scintilla etc. or the backup code well. The patch *seems* to work against some initial testing I've done, but the patch is as much to demonstrate the problem and example solution to that.

However the file opening logic in bowpad is complicated and really needs to be refactored so it is more maintainable and less fragile. Recent support for backups has made the code even more complicated. Personally, I wonder if the backup feature should be dropped to keep things simpler and reintroduced later under a better model or something. Or even not, as I find work using the backup files a questionable work flow anyway.

I think pondering how this file opening code works in more scenarios is necessary and with a view to a simpler model. But for now this patch offers something to play with that demonstrates the issue and a possible solution to test with.

### CODE REVIEW

File: src/Commands/CmdSession.cpp [hunk 1]
```diff
@@ -31,6 +31,8 @@
 #include "ProgressDlg.h"
 #include "SysInfo.h"
 #include "AppUtils.h"
+#include "MainWindow.h"
+
```
Comment (stefankueng) [inline src/Commands/CmdSession.cpp:35]: commands should not have access to the main window.
Instead, extend the OpenFile() method of the ICommand interface.

File: src/MainWindow.cpp [hunk 2]
```diff
@@ -3165,145 +3165,159 @@ int CMainWindow::OpenFile(const std::wstring& file, unsigned int openFlags)
...
-            if (m_fileTree.GetPath().empty())
-            {
-                m_fileTree.SetPath(CPathUtils::GetParentDirectory(filepath), false);
+        // only activate the new doc tab if the main window is enabled:
+        // if it's disabled, a modal dialog is shown
+        // (e.g., the handle-outside-modifications confirmation dialog)
+        // and we then must not change the active tab.
+        // SCI_SETDOCPOINTER is necessary so the reference count of the document
+        // is decreased and the memory can be released.
+        if (IsWindowEnabled(*this))
```
Comment (stefankueng) [inline src/MainWindow.cpp:3293]: the comment here clearly states that SCI_SETDOCPOINTER is necessary, but then you removed the call?

File: src/ScintillaWnd.cpp [hunk 3]
```diff
@@ -829,43 +826,32 @@ void CScintillaWnd::UpdateLineNumberWidth()
 
 void CScintillaWnd::SaveCurrentPos(CPosData& pos)
 {
```
Comment (stefankueng) [inline src/ScintillaWnd.cpp:828]: again, the comment clearly states to NOT save the positions. Why did you remove that?

File: src/ScintillaWnd.cpp [hunk 4]
```diff
@@ -2410,20 +2379,6 @@ void CScintillaWnd::ReflectEvents(SCNotification* pScn)
...
-                Call(SCI_SETFIRSTVISIBLELINE, visLineToScrollTo);
-                --m_LineToScrollToAfterPaintCounter;
-                if ((m_LineToScrollToAfterPaintCounter <= 0) || ((Call(SCI_GETFIRSTVISIBLELINE) + 2) > Call(SCI_VISIBLEFROMDOCLINE, m_LineToScrollToAfterPaint)))
-                {
-                    m_LineToScrollToAfterPaint        = -1;
-                    m_WrapOffsetToScrollToAfterPaint  = 0;
-                    m_LineToScrollToAfterPaintCounter = 0;
-                }
-                UpdateLineNumberWidth();
-            }
```
Comment (stefankueng) [inline src/ScintillaWnd.cpp:2426]: When wrapping is enabled, all the "go to line" commands in Scintilla won't work properly because the wrapping info is only calculated when the scintilla view is painted. That's why this code is here.
Why did you remove this?

File: src/ScintillaWnd.cpp [hunk 5]
```diff
@@ -913,18 +899,13 @@ void CScintillaWnd::RestoreCurrentPos(const CPosData& pos)
     }
     Call(SCI_CHOOSECARETX);
 
+    Call(SCI_ENSUREVISIBLE, pos.m_nFirstVisibleLine);
```
Comment (stefankueng) [inline src/ScintillaWnd.cpp:902]: here, I get a complete hang: scintilla loops indefinitely on the fourth restored file (of 9 files to open on startup)
The problem is that SCI_ENSUREVISIBLE requires the view to paint, but that doesn't happen.

Comment (gmcode) [reply src/ScintillaWnd.cpp:902]: I'm not expert enough with scintilla to understand this, but it doesn't surprise me it would be an affect which is why my comment says it needs testing thoroughly. Hopefully you can fix this particular issue without losing the fixes my patch aims to address.

File: src/ScintillaWnd.cpp (hunk 4 above)
Comment (gmcode) [reply src/ScintillaWnd.cpp:2426]: I removed the code because I am *hoping* the code I removed is working around a problem the old code created and which I am *hoping* the new code fixes. I'm not seeing any wrapping problems yet, not that It would surprise me as I don't claim to be expert here as my comment says. What new problem do you see here related to wrapping?

File: src/ScintillaWnd.cpp (hunk 3 above)
Comment (gmcode) [reply src/ScintillaWnd.cpp:828]: Because I'm not sure the comment is correct any longer. See below.

File: src/MainWindow.cpp (hunk 2 above)
Comment (gmcode) [reply src/MainWindow.cpp:3293]: I removed the SCI_SETDOCPOINTER call, because it was never ever called, it was only called if bActivate was false and it was always true.

File: src/ScintillaWnd.cpp (hunk 5 above)
Comment (stefankueng) [reply src/ScintillaWnd.cpp:902]: the problem here is that you "fixed" your flicker problem by forcing Scintilla to paint every document in full right after loading it. This causes an immense slowdown when you have bigger files or complex lexers.
So I'm sorry to say that this is not a fix I can approve - the slowdown is just too big (waiting 5 minutes is really unacceptable, and that's just for one big logfile - imagine I had more logfiles open or even my 300MB generated c++ file which has a much more complex lexer)

Comment (gmcode) [reply src/ScintillaWnd.cpp:902]: You're making it sound like bowpad must draw the screen at page 1 and then again at the right place in order to prevent the slowdown. If that was actually the case every editor that used scintilla would have this problem and they don't, so the problem must be fixable.

By far the bulk of my patch is to ensure the position information is available to Bowpad early enough to use at the right time. So that means you can take everything in my patch except the changes to ScntillaWnd.cpp. So I suggest you take all of the files minus ScntillaWnd.cpp, and then patch the previous version of ScintillaWnd.cpp prior to my version, to your liking, so that it uses the position information that's now available to it to avoid the problem. That shouldn't be much work for you to finish it off.  There's no way you have to draw the screen twice to do this. What's left for you to do should be a small change for you, maybe one line or two of code now.

Comment (gmcode) [reply src/ScintillaWnd.cpp:902]: Oh I forgot to say: can you please confirm you actually tested your slowdown test case on the version I've submitted and that your slowdown comments isn't just based on your memory of your prior code. I just want to be sure it still applies. But whether you patch my ScntillaWnd.cpp or the prior version, I'm sure there can't be much left to fix the drawing problem now the location information is available at the right time. Check the comment in the Centre function, you can perhaps conditionally call SCI_ENSUREVISIBLE like it does to avoid the extra call. That might be enough. Whatever the issue is that's remaining, I'm sure it isn't much.

Comment (gmcode) [reply src/ScintillaWnd.cpp:902]: It might be worth mentioning what your slowdown test case is too so I can try it on my machine. I don't really doubt there's an issue there, just that fixing it is a small patch to my patch if anything and you're better placed than me to know what that change is.

Comment (stefankueng) [reply src/ScintillaWnd.cpp:902]: just use a very big file that needs a lexer. Around 500MB big should suffice to show you a significant delay when loading that file on startup (i.e. using a session).
I have here a log file (*.log) that's 900MB big.
And yes, I've tested this, otherwise I wouldn't have noticed the delay/slowdown.

Comment (gmcode) [reply src/ScintillaWnd.cpp]: I created a 300mb test file by appending src\Commands\*.cpp to the test file other over and over until the test file size was 300mb.  The test file contained 7.6 million lines of c++.

I then compared the time taken to load that test file by the last released version of bowpad, against the time taken to load that same file with a version of bowpad patched with my original patch in this pull request.

The tl;dr is: my additional one line patch makes both versions load at the same time of about 1.45 minutes. So unless you find anything conceptually wrong with it, please apply this change. Without this second patch one line patch,  the version without the extra 1 line patch takes about 30 seconds longer.

File: src/Commands/CmdSession.cpp (hunk 1 above)
Comment (gmcode) [reply src/Commands/CmdSession.cpp:34]: This was an oversight. Fixed.

File: src/ScintillaWnd.cpp (hunk 5 above)
Comment (stefankueng) [reply src/ScintillaWnd.cpp]: I still get a slowdown, not as much but still noticeable. But I can overlook that.
However, #134 was fixed and with this PR it stops working again.

Comment (gmcode) [reply src/ScintillaWnd.cpp]: Hmm, I explicitly tested that situation and was sure I found it worked. But you're right, I just tested it again and it's back to not working. I'll take a look again.

**Comment by gmcode** [comment]
It seems the hopeful performance fix patch broke the previous commit. So I've changed the code back to working closer to the first commit and made more changes to fix other issues I found in position restoration during testing.

So now restoring the position seems to work in all cases that I tested. So:

The main pros of this patch branch are:
* It fixes the issue of painting page 1 and then repainting it.
* It removes a tone of code that really feels like workaround code that is working around a problem of bowpad's own creation.

The main cons of the patch are::
* It definitely affects performance and I don't know how to fix it in bowpad, but I think that's not the place to fix it.
* Indent wrapping seems to paint late, but I think it can be fixed, and that's not as annoying as drawing a whole wrong page wrongly and then redrawing everything again, which is what the code this patch branch is attempting to fix did. This seems a small problem that can be solved, but we'll have to look at this together.
 
So here's what I think overall:

* The late wrapping thing can probably be fixed, so can you have a think about why that's happening there please. If we can fix indent wrapping simply, then can remove a lot of workaround code that might only get bigger because we then don't have bowpad working against scintilla which I think is currently the case and bowpad drawing whole wrong pages.

If we can do that, we'd have all that workaround code removed, which is a lot cleaner and fixed the double painting problem, which together is a good win, which would just leave the performance issue.

And on the performance subject:

Looking at the wrapping performance issue, it seems to me that THAT issue is primarily Scintilla's issue. I stepped through Scintilla's wrapping code and saw bowpad was spending most of it's time there. It probably needs to get its memory allocation management tuned there.

I say this also because by comparison, I opened up a 300mb file in Visual Studio Code and noticed it behaved MUCH better. First it suggested disabling wrapping so things didn't lock up, exactly like what Bowpad does and always has. It then offers to let you proceed disabling or enable wrapping in the process. If you disable wrapping, code loads instantly. If you enable wrapping, Visual Studio Code opens the the 300mb file in about 10 seconds.

That tells me a) bowpad could offer similar open with/without wrapping options to get performance for the worst cases, and b) it is Scintilla that likely needs to change to get the performance back and isn't anything Bowpad should really do in that case, because the performance changes in Scintilla would likely be massive performance improvements, not minor ones as tweaking bowpad would likely get.

I also think performance is second to correctness given that the performance issues affect massive file and file(s) use cases, so favoring a niche set of users over a few large file (millions of lines) cases doesn't make sense to me and especially when massive performance gains ikely needs to come from Scintilla and changes there would massively benefit those niche users.

So my view now is that it would help things progress if we use the patches in this branch as the baseline, especially if we can fix what the late indent wrap painting issue and move on from there and perhaps you could suggest to the scintilla people they might have a problem there. I could be wrong in all of this but that's my view until someone else investigates my suggestions.

Also I have other fixes (like to the horizontal scrolling problem) that require this patch to be in place and if it's not I'm going to have difficulty with git and likely give up.

**Comment by gmcode** [comment]
I've discovered the late indent wrapping only occurs when there is more than one tab open, so I should be able to find this remaining problem soon.

**Comment by gmcode** [comment]
The wrapping commands need to load before the session, otherwise the session activates the documents in the session but they inherit the wrong settings. This patch introduces a load first list (before the session in this case) to avoid that. Other settings probably want to go in this list too, but that's beyond the scope of what this particular patch is trying to address right now.

This commit concludes all positioning/painting problems I intend to fix with this commit branch. 

To see the problem this patch fixes, open bowpad without this last commit, with autoload on, then load MainWindow.cpp  and ctrl-end to the end, then open any any other file, then go the first tab and exit. Then reload, you should notice then bowpad draws the page but then does the indentation as a late second action. With this patch applied, bowpad will now draw the page successfully in one operation.

The test case with this branch applied should draw demonstrably cleaner than released versions of bowpad.

This patch is ready for consideration for applying now.

**Comment by stefankueng** [comment]
and now the slowdown is here again.
and no, performance should **never** be considered second to UI glitches. Also, even though wrapping performance is a Scintilla problem, that also makes it a BP problem. Just hoping that Scintilla will improve here someday is not an option.
